Unable to compile certain bevy crates directly, unless an executor is defined separately for `bevy_tasks`

Hi! While writing some of my linting PRs, I seem to have come across an odd problem with `bevy_tasks`.

Specifically, it seems that certain crates rely on `bevy_tasks`, but then prevent `bevy_tasks` from having any executor features enabled, causing certain executor types to not be defined.

I was able to observe this while trying to compile the `bevy_state` and `bevy_hierarchy` crates directly (using `--package` to select them specifically), but it's possible there may be other bevy crates that do the same.

<details><summary>Full compilation error log</summary>

```Shell
error[E0412]: cannot find type `ExecutorInner` in this scope
  --> crates\bevy_tasks\src\executor.rs:39:25
   |
39 | pub struct Executor<'a>(ExecutorInner<'a>);
   |                         ^^^^^^^^^^^^^ not found in this scope

error[E0412]: cannot find type `LocalExecutorInner` in this scope
  --> crates\bevy_tasks\src\executor.rs:49:30
   |
49 | pub struct LocalExecutor<'a>(LocalExecutorInner<'a>);
   |                              ^^^^^^^^^^^^^^^^^^ not found in this scope

error[E0599]: the method `get` exists for struct `EagerStorage<LocalExecutor<'static>>`, but its trait bounds were not satisfied
  --> crates\bevy_tasks\src\single_threaded_task_pool.rs:22:1
   |
22 | / thread_local! {
23 | |     static LOCAL_EXECUTOR: LocalExecutor<'static> = const { LocalExecutor::new() };
24 | | }
   | |_^ method cannot be called on `EagerStorage<LocalExecutor<'static>>` due to unsatisfied trait bounds
   |
  ::: crates\bevy_tasks\src\executor.rs:49:1
   |
49 |   pub struct LocalExecutor<'a>(LocalExecutorInner<'a>);
   |   ---------------------------- doesn't satisfy `_: Sized`
   |
   = note: the following trait bounds were not satisfied:
           `{type error}: core::marker::Sized`
           which is required by `executor::LocalExecutor<'static>: core::marker::Sized`
   = note: this error originates in the macro `$crate::thread::local_impl::thread_local_inner` which comes from the expansion of the macro `thread_local` (in Nightly builds, run with -Z macro-backtrace for mor
error[E0433]: failed to resolve: use of undeclared type `ExecutorInner`
  --> crates\bevy_tasks\src\executor.rs:56:14
   |
56 |         Self(ExecutorInner::new())
   |              ^^^^^^^^^^^^^ use of undeclared type `ExecutorInner`

error[E0433]: failed to resolve: use of undeclared type `LocalExecutorInner`
  --> crates\bevy_tasks\src\executor.rs:65:14
   |
65 |         Self(LocalExecutorInner::new())
   |              ^^^^^^^^^^^^^^^^^^
   |              |
   |              use of undeclared type `LocalExecutorInner`
   |              help: a struct with a similar name exists: `LocalExecutor`

error[E0599]: the method `with` exists for struct `LocalKey<LocalExecutor<'static>>`, but its trait bounds were not satisfied
   --> crates\bevy_tasks\src\single_threaded_task_pool.rs:208:31
    |
208 |         return LOCAL_EXECUTOR.with(|executor| {
    |                ---------------^^^^ method cannot be called on `LocalKey<LocalExecutor<'static>>` due to unsatisfied trait bounds
    |
   ::: crates\bevy_tasks\src\executor.rs:49:1
    |
49  | pub struct LocalExecutor<'a>(LocalExecutorInner<'a>);
    | ---------------------------- doesn't satisfy `_: Sized`
    |
    = note: the following trait bounds were not satisfied:
            `{type error}: core::marker::Sized`
            which is required by `executor::LocalExecutor<'static>: core::marker::Sized`

error[E0599]: the method `with` exists for struct `LocalKey<LocalExecutor<'static>>`, but its trait bounds were not satisfied
   --> crates\bevy_tasks\src\single_threaded_task_pool.rs:253:31
    |
253 |         return LOCAL_EXECUTOR.with(f);
    |                               ^^^^ method cannot be called on `LocalKey<LocalExecutor<'static>>` due to unsatisfied trait bounds
    |
   ::: crates\bevy_tasks\src\executor.rs:49:1
    |
49  | pub struct LocalExecutor<'a>(LocalExecutorInner<'a>);
    | ---------------------------- doesn't satisfy `_: Sized`
    |
    = note: the following trait bounds were not satisfied:
            `{type error}: core::marker::Sized`
            which is required by `executor::LocalExecutor<'static>: core::marker::Sized`

Some errors have detailed explanations: E0412, E0433, E0599.
For more information about an error, try `rustc --explain E0412`.
error: could not compile `bevy_tasks` (lib) due to 7 previous errors
```
</details>

I would make a PR for this, but I'm unsure how to handle it.

--- a/crates/bevy_tasks/src/lib.rs
+++ b/crates/bevy_tasks/src/lib.rs
@@ -16,6 +16,9 @@ extern crate std;
 
 extern crate alloc;
 
+#[cfg(not(any(feature = "async_executor", feature = "edge_executor")))]
+compile_error!("Either of the `async_executor` or the `edge_executor` features must be enabled.");
+
 #[cfg(not(target_arch = "wasm32"))]
 mod conditional_send {
     /// Use [`ConditionalSend`] to mark an optional Send trait bound. Useful as on certain platforms (eg. Wasm),
@@ -45,6 +48,7 @@ pub type BoxedFuture<'a, T> = core::pin::Pin<Box<dyn ConditionalSendFuture<Outpu
 
 pub mod futures;
 
+#[cfg(any(feature = "async_executor", feature = "edge_executor"))]
 mod executor;
 
 mod slice;

--- a/crates/bevy_tasks/src/single_threaded_task_pool.rs
+++ b/crates/bevy_tasks/src/single_threaded_task_pool.rs
@@ -12,11 +12,49 @@ use portable_atomic_util::Arc;
 #[cfg(not(feature = "portable-atomic"))]
 use alloc::sync::Arc;
 
-#[cfg(feature = "std")]
+#[cfg(all(
+    feature = "std",
+    any(feature = "async_executor", feature = "edge_executor")
+))]
 use crate::executor::LocalExecutor;
 
-#[cfg(not(feature = "std"))]
+#[cfg(all(
+    not(feature = "std"),
+    any(feature = "async_executor", feature = "edge_executor")
+))]
 use crate::executor::Executor as LocalExecutor;
+
+#[cfg(not(any(feature = "async_executor", feature = "edge_executor")))]
+mod dummy_executor {
+    use async_task::Task;
+    use core::{future::Future, marker::PhantomData};
+
+    /// Dummy implementation of a `LocalExecutor` to allow for a cleaner compiler error
+    /// due to missing feature flags.
+    #[doc(hidden)]
+    #[derive(Debug)]
+    pub struct LocalExecutor<'a>(PhantomData<fn(&'a ())>);
+
+    impl<'a> LocalExecutor<'a> {
+        /// Dummy implementation
+        pub const fn new() -> Self {
+            Self(PhantomData)
+        }
+
+        /// Dummy implementation
+        pub fn try_tick(&self) -> bool {
+            unimplemented!()
+        }
+
+        /// Dummy implementation
+        pub fn spawn<T: 'a>(&self, _: impl Future<Output = T> + 'a) -> Task<T> {
+            unimplemented!()
+        }
+    }
+}
+
+#[cfg(not(any(feature = "async_executor", feature = "edge_executor")))]
+use dummy_executor::LocalExecutor;
 
 #[cfg(feature = "std")]
 thread_local! {

